File: 2026/2205/2205d.cpp
/*!
 * author: manhhomienbienthuy
 * created: 2026-07-06T01:52:29+00:00
 * https://codeforces.com/contest/2205/problem/D
 */

#include <bits/stdc++.h>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int t;
  cin >> t;
  while (t--) {
    int n;
    cin >> n;
    vector<int> a(n);
    for (auto& x : a) cin >> x;

    vector<int> parent(n, -1), left(n, -1), right(n, -1);
    stack<int> st;
    int root = -1;
    for (int i = 0; i < n; i++) {
      int last = -1;
      while (!st.empty() && a[st.top()] < a[i]) {
        last = st.top();
        st.pop();
      }

      if (st.empty()) {
        root = i;
      } else {
        parent[i] = st.top();
        right[st.top()] = i;
      }

      if (last != -1) {
        parent[last] = i;
        left[i] = last;
      }

      st.push(i);
    }

    vector<int> q = {root};
    for (int i = 0; i < (int)q.size(); i++) {
      int u = q[i];
      if (left[u] != -1) q.push_back(left[u]);
      if (right[u] != -1) q.push_back(right[u]);
    }
    reverse(q.begin(), q.end());

    vector<int> sz(n, 0), dp(n, 0);
    for (auto u : q) {
      int sz_left = left[u] == -1 ? 0 : sz[left[u]];
      int sz_right = right[u] == -1 ? 0 : sz[right[u]];
      sz[u] = 1 + sz_left + sz_right;

      int dp_left = left[u] == -1 ? 0 : dp[left[u]];
      int dp_right = right[u] == -1 ? 0 : dp[right[u]];
      dp[u] = min(sz_right + dp_left, sz_left + dp_right);
    }

    cout << dp[root] << '\n';
  }

  return 0;
}

File: 2026/2205/2205d.py
t = int(input())
for _ in range(t):
    n = int(input())
    a = list(map(int, input().split()))

    parent = [-1] * n
    left = [-1] * n
    right = [-1] * n

    stack = []
    root = -1
    for i in range(n):
        last = -1
        while stack and a[stack[-1]] < a[i]:
            last = stack.pop()

        if stack:
            right[stack[-1]] = i
            parent[i] = stack[-1]
        else:
            root = i

        if last != -1:
            left[i] = last
            parent[last] = i
        stack.append(i)

    q = [root]
    for u in q:
        if left[u] != -1:
            q.append(left[u])
        if right[u] != -1:
            q.append(right[u])

    sz = [0] * n
    dp = [0] * n

    for u in reversed(q):
        ls = sz[left[u]] if left[u] != -1 else 0
        rs = sz[right[u]] if right[u] != -1 else 0
        left_dp = dp[left[u]] if left[u] != -1 else 0
        right_dp = dp[right[u]] if right[u] != -1 else 0
        sz[u] = 1 + ls + rs
        dp[u] = min(
            rs + left_dp,
            ls + right_dp
        )

    print(dp[root])
